### DISCUSSION TITLE
Password Input

### DISCUSSION THREAD

**Discussion by Rozerinay** [nl-design-system/backlog#285] [Component suggestions]
<img width="480" alt="visuele weergave van een password input component" src="https://github.com/user-attachments/assets/1fb69bb5-0bff-4afb-8552-516b4a23df44">

We willen graag meer leren over de Password Input component. In welke vorm, of vormen, wordt deze momenteel ingezet en toegepast binnen jouw organisatie?

Deel hieronder een link naar Figma, Storybook, Github of screenshots. Met deze informatie kunnen we bewijzen dat de component onderdeel moet worden van de NL Design System component library.

Daarnaast wordt het delen van onderzoek gewaardeerd. Dus heb je zelf onderzoek gedaan en zijn er inzichten voor dit component? Dan horen we dit ook graag!

[Link naar bericht in Slack](https://github.com/orgs/nl-design-system/discussions/285)

---

## Naam
Password Input

## Doel
Afgeschermd invoerveld om een geheime tekst in te vullen, zoals een wachtwoord.

## Gerelateerde componenten
[Checkbox Group](https://github.com/orgs/nl-design-system/discussions/223), [Date Input](https://github.com/orgs/nl-design-system/discussions/270), [File Input](https://github.com/orgs/nl-design-system/discussions/310), [Number Input](https://github.com/orgs/nl-design-system/discussions/235), [Radio Group](https://github.com/orgs/nl-design-system/discussions/222), [Text Area](https://github.com/orgs/nl-design-system/discussions/297), [Text Input](https://github.com/orgs/nl-design-system/discussions/225), [Text Input Group](https://github.com/orgs/nl-design-system/discussions/296)

## Onderzoek
Ik heb in een eerder stadium al eens onderzoek gedaan naar de Password Input component ([te vinden in het Figma Schetsboek](https://www.figma.com/design/fy08SZpZmqx6ljLwvA3Woe/NLDS---Schetsboek?node-id=1039-6351&t=MJUFjPL08n19trjz-4)). Hieronder staan notities die ik toen maakte. Hier kunnen notities tussen staan die eerder betrekking hebben tot een patroon rondom 'Aanmelden' of 'Inloggen'. Maar omdat er overlap in zit bewaar ik de notities hier.

**Implementatie**

[`<input type="password">` - MDN docs](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/password)
Documentatie van Mozilla Developers Network over het `<input>` HTML element van met  `type="password"`.

Er is een verschilt tussen een wachtwoord aanmaken (Aanmelden) en wachtwoord invullen (Inloggen). Hier kunnen ook twee verschillende auto-complete ... voor gebruikt worden:
  - Aanmelden: `autocomplete=”new-password”` [MDN docs - New Password](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/password#new-password)
  - Inloggen: `autocomplete="current-password"` [MDN docs - Current Password](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/password#current-password)

**Wat kunnen we leren van GOV.UK?**

[Password Input - GOV.UK](https://design-system.service.gov.uk/components/password-input/)
Password Input implementatie bij GOV.UK. Onder 'Research on this component' worden bevindingen genoteerd. Er is ook een [link naar de Pasword Input GitHub Discussion](https://github.com/alphagov/govuk-design-system-backlog/issues/240) waar meer informatie, onderzoek en aanpassingen worden gedeeld. Enkele interessante comments:

- [Geen 2e invulveld](https://github.com/alphagov/govuk-design-system-backlog/issues/240#issuecomment-2020125340). In deze comment wordt uitgelegt waarom ze zijn afgestapt van het inzetten van een 2e invulveld om men een wachtwoord te laten bevestigen.
- [Spellcheck = false](https://github.com/alphagov/govuk-design-system-backlog/issues/56#issuecomment-1251669827). In deze comment wordt verwezen naar een blogpost die helaas niet meer bestaat. Maar de strekking is dat het `spellcheck` attribute `false` zou moeten zijn waardoor het wachtwoord niet met een externe partij zal worden gedeeld. Ook niet wanneer het wachtwoord getoond wordt door middel van een wijziging naar `<input type = "text">`.
- `autocapitalize="none"`?
- `autocorrect="off"`?
- [Icons](https://github.com/alphagov/govuk-design-system-backlog/issues/240#issuecomment-1329394063). In deze comment wordt aangegeven dat het verstandig is rekening te houden met password managers en Safari. Dit is ook relevant voor de positionering van een toon/verberg wachtwoord functionaliteit.
- [Toon / Verberg functie helpt toegankelijkheid](https://github.com/alphagov/govuk-design-system-backlog/issues/240#issue-1032347551). In deze comment wordt aangegeven dat een Toon / Verberg functionaliteit kan helpen als iemand moeite heeft met typen.
- [Wachtwoord direct als tekst tonen](https://github.com/alphagov/govuk-design-system-backlog/issues/56#issuecomment-456900147). In deze comment wordt aangegeven dat het mensen kan afschrikken wanneer je een wachtwoord direct als tekst toont. Maar dat een Toon / Verberg functionaliteit wel wordt gewaardeerd.

**Voorkom fouten**

[8 UI/UX tips about password design - DEV Community](https://dev.to/indieklem/8-uiux-tips-about-password-design-5bbn#:~:text=2.%20Show%20if%20the%20Caps%20Lock%20key%20is%20on)
Artikel met tips rondom het ontwerpen van een wachtwoord veld. 1 tip gaat over geven van feedforward dat caps of num loc aanstaat.

**Wachtwoord criteria**

[Avoid Unnecessarily Complex Password-Creation Requirements – Baymard Institute](https://baymard.com/blog/password-requirements-and-password-reset)
Artikel van Baymard Institute waarin wordt aangeraden om wachtwoord criteria niet te lastig maken.

[Password Creation: 3 Ways To Make It Easier - NN Group](https://www.nngroup.com/articles/password-creation/)
Artikel van NN Group waarin wordt onder andere wordt aangeraden om:
- een indicatie te geven hoe 'sterk' het wachtwoord is.
- wachtwoord criteria direct te tonen.

Vraag: Zou het helpen als wachtwoord criteria niet enkel bij 'Aanmelden' wordt getoond maar ook bij 'Inloggen'? [Zie schets](https://www.figma.com/design/fy08SZpZmqx6ljLwvA3Woe/NLDS---Schetsboek?node-id=1039-6362&t=MJUFjPL08n19trjz-4).

[Designing a better password input field - Medium](https://medium.com/@flik185/designing-a-better-password-input-field-140446e0ed08)
Artikel waarin wordt voorgesteld om direct feedback te geven als men aan wachtwoord regel voldoet. [GOV.UK is hier nog niet uit](https://design-system.service.gov.uk/patterns/passwords/#:~:text=guidance%20on%20passwords-,Next%20steps,-More%20research%20is)

**Toon / Verberg functionaliteit**

[Simple things are complicated: making a show password option - GOV.UK](https://technology.blog.gov.uk/2021/04/19/simple-things-are-complicated-making-a-show-password-option/)
Artikel van Andy Sellick waarin hij vertelt over het implementeren van een Toon / Verberg functionaliteit voor GOV.UK. En de lessen die ze daar bij leerden.

[Show/Hide password accessibility and password hints tutorial - Part of a Whole](https://incl.ca/show-hide-password-accessibility-and-password-hints-tutorial/)
Artikel van Nic Steenhout waarin het toegankelijk opzetten van een Toon / Verberg functionaliteit wordt beschreven.

**Comment by jeffreylauwers** [comment]
## DSO

Geen component kunnen vinden voor Password Input maar in de [documentatie over 'Inline validatie'](https://www.dso-toolkit.nl/62.8.4/components/form/#inline-validatie) wordt er wel iets over geschreven.

> Valideer niet na elk getypt teken. Deze aanpak verhoogt niet alleen het aantal onnodige validatiepogingen, maar frustreert gebruikers ook. Er zijn een paar uitzonderingen op deze regel:
> - Het is handig om inline te valideren terwijl de gebruiker typt bij het maken van een wachtwoord. Zo controleert u of het wachtwoord voldoet aan de eisen.

**Reply by tfrijsewijk** [reply]
We documenteren het niet, maar de DSO Toolkit heeft een component "Input Password" (what's in a name). We etaleren dit component als onderdeel van "Form Group: Input".

* https://storybook.dso-toolkit.nl/65.0.2/?path=/docs/html-css-form-groups-input--docs
* https://storybook.dso-toolkit.nl/65.0.2/?path=/story/html-css-form-groups-input--input&args=type:password

Form Groups worden opgebouwd uit andere componenten (Label, Input, Info) en bevat ook eigen onderdelen (de validatie foutmelding, feedback icons, help tekst) en mogen _alleen_ in een Form worden gebruikt.

**Reply by tfrijsewijk** [reply]
Verder zie ik dat je linkt naar een oude versie (62.8.4). Werk je toevallig met een bookmark naar een fixed version? Beter kan je dso-toolkit.nl openen/bookmarken, dan krijg je een redirect naar de laatste versie.

**Reply by jeffreylauwers** [reply]
Ah, ja dat klopt Thomas. Ik heb de 'bookmark' aangepast. 👍

**Comment by jeffreylauwers** [comment]
## DUO

Geen component kunnen vinden voor Password Input maar de component wordt wel ingezet voor [het patroon 'Inloggen'](https://uno.dfront.rijkscloud.nl/#/patronen/inloggen-pagina).

![Screenshot inlogscherm bij DUO met velden voor e-mailadres en wachtwoord](https://github.com/user-attachments/assets/6f4702b4-bd66-4d12-8e9d-dc71b0e072c6)

**Comment by jeffreylauwers** [comment]
## OpenGemeenten

[Patternlab - Forms & Search](https://patternlab.lenteveld.nl/patterns/01-molecules-080-Forms-And-Search/index.html)

---

### Screenshots / Varianten

**Patternlab**

Default
![Scherm­afbeelding 2024-10-07 om 15 10 48](https://github.com/user-attachments/assets/adf1c4f3-d136-4ca2-9462-381e75404f09)


[AdvancedPassword Description](https://patternlab.lenteveld.nl/patterns/01-molecules-080-Forms-And-Search-AdvancedPassword-AdvancedPassword-Description/01-molecules-080-Forms-And-Search-AdvancedPassword-AdvancedPassword-Description.html)
![Scherm­afbeelding 2024-10-07 om 15 11 20](https://github.com/user-attachments/assets/89926195-86fd-45e6-a207-9699ae5e271c)


[AdvancedPassword Error](https://patternlab.lenteveld.nl/patterns/01-molecules-080-Forms-And-Search-AdvancedPassword-AdvancedPassword-Error/01-molecules-080-Forms-And-Search-AdvancedPassword-AdvancedPassword-Error.html)
![Scherm­afbeelding 2024-10-07 om 15 12 30](https://github.com/user-attachments/assets/9d91a544-c257-4143-9327-ebf0137ccc73)


[AdvancedPassword Placeholder](https://patternlab.lenteveld.nl/patterns/01-molecules-080-Forms-And-Search-AdvancedPassword-AdvancedPassword-Placeholder/01-molecules-080-Forms-And-Search-AdvancedPassword-AdvancedPassword-Placeholder.html)
![Scherm­afbeelding 2024-10-07 om 15 12 36](https://github.com/user-attachments/assets/5f3f1b29-3661-4ff8-9d3f-e8ae88f92685)

[AdvancedPassword Required](https://patternlab.lenteveld.nl/patterns/01-molecules-080-Forms-And-Search-AdvancedPassword-AdvancedPassword-Required/01-molecules-080-Forms-And-Search-AdvancedPassword-AdvancedPassword-Required.html)
![Scherm­afbeelding 2024-10-07 om 15 12 41](https://github.com/user-attachments/assets/286045e5-4ce3-4352-881a-1668e6a51121)

[AdvancedPassword Value](https://patternlab.lenteveld.nl/patterns/01-molecules-080-Forms-And-Search-AdvancedPassword-AdvancedPassword-Value/01-molecules-080-Forms-And-Search-AdvancedPassword-AdvancedPassword-Value.html)
![Scherm­afbeelding 2024-10-07 om 15 12 45](https://github.com/user-attachments/assets/8f7b2f14-6aaa-4b96-9285-b4561f985a72)

Voorbeeld van 1Password icoon bij focus
![Scherm­afbeelding 2024-10-07 om 15 12 52](https://github.com/user-attachments/assets/7399136a-7336-487e-9607-c22747ec097b)

Voorbeeld van tooltip die getoond wordt wanneer men over 1Password icoon hovert.
![Scherm­afbeelding 2024-10-07 om 15 13 01](https://github.com/user-attachments/assets/0604cca3-112d-47a9-aa88-f3adfdeda511)

**Comment by jeffreylauwers** [comment]
## Utrecht

Als onderdeel van Textbox component.

- [Storybook - Textbox --> Wachtwoord](https://nl-design-system.github.io/utrecht/storybook/?path=/docs/css_css-textbox--docs#wachtwoord)
- [Storybook - Textbox --> Wachtwoord kiezen](https://nl-design-system.github.io/utrecht/storybook/?path=/docs/css_css-textbox--docs#wachtwoord-kiezen)
- [GitHub](https://github.com/nl-design-system/utrecht/tree/main/components/textbox)

---

### Screenshots / Varianten

**Storybook**

Wachtwoord 
Het is belangrijk dat gebruikers een password manager kunnen gebruiken om een wachtwoord in te vullen ([WCAG eis 3.3.7](https://www.w3.org/TR/WCAG22/#accessible-authentication)). Gebruik nooit `autocomplete="off"` voor een wachtwoordveld. Gebruikers moeten een wachtwoord uit een password manager kunnen plakken, gebruik daarom nooit code om kopiëren en plakken in dit veld te blokkeren. Gebruik `spellCheck="false"` zodat het wachtwoord niet wordt gedeeld met browser extensions voor spellingcontrole, en zodat een wachtwoord met opzettelijke spelfouten niet wordt gecorrigeerd.
![Scherm­afbeelding 2024-10-07 om 15 38 20](https://github.com/user-attachments/assets/71b52744-51d2-4613-8ee8-b357f6300ef2)


Wachtwoord kiezen
Het is belangrijk dat gebruikers een password manager kunnen gebruiken om nieuw wachtwoord kiezen en te onthouden ([WCAG eis 3.3.7](https://www.w3.org/TR/WCAG22/#accessible-authentication)). Gebruik nooit `autocomplete="off"` voor een wachtwoordveld. Gebruikers moeten een nieuw wachtwoord uit een password manager kunnen plakken, gebruik daarom nooit code om kopiëren en plakken in dit veld te blokkeren. Gebruik `spellcheck="false"` zodat het wachtwoord niet wordt gedeeld met browser extensions voor spellingcontrole, en zodat een wachtwoord met opzettelijke spelfouten niet wordt gecorrigeerd.
![Scherm­afbeelding 2024-10-07 om 15 38 20](https://github.com/user-attachments/assets/cb57c1ca-efac-4c04-818c-fadf5b496c9d)

---

### Notities

- Als onderdeel van Textbox component.
- Documentatie beschikbaar met handige tips voor implementatie en verwijzing naar WCAG richtlijn.

**Comment by jeffreylauwers** [comment]
## ✨ Dit component is nu Help Wanted ✨

Help je mee hem door de Community stappen te krijgen?
[Password Input op het Community bord](https://github.com/orgs/nl-design-system/projects/29/views/1?filterQuery=password+input)

**Comment by eslook** [comment]
## ✨ Dit component is nu Community ✨

Help je mee hem door de Candidate stappen te krijgen?
[Password Input op het Candidate bord](https://github.com/orgs/nl-design-system/projects/80/views/1?filterQuery=Password+Input)
